### PR TITLE
feat: complete Batch 2 assistant payload compaction

### PR DESCRIPTION
**Pull request by thienchi2109**
## Summary
- complete Batch 2 of `refactor-assistant-tool-payload-compaction`
- migrate `categorySuggestion` to bounded candidate retrieval via `ai_category_suggestion` with tool-schema `device_name` mapped to RPC param `p_device_name`
- migrate `departmentList` to the pass-1 envelope contract without `uiArtifact`, keep model-visible department count data, and update prompt guidance for reading envelope outputs
- add route-level whitelist coverage for `ai_category_suggestion` and `ai_department_list`
- update OpenSpec Batch 2 checklist and add the implementation plan document

## Verification
- `node scripts/npm-run.js run verify:no-explicit-any`
- `node scripts/npm-run.js run typecheck`
- `node scripts/npm-run.js run test:run -- src/lib/ai/tools/__tests__/category-suggestion-envelope.test.ts src/lib/ai/tools/__tests__/category-suggestion-registry.test.ts src/lib/ai/tools/__tests__/department-list-envelope.test.ts src/lib/ai/prompts/__tests__/system-prompt-envelope.test.ts src/lib/ai/prompts/__tests__/system.test.ts src/app/api/chat/__tests__/route.tool-rpc-mapping.test.ts src/app/api/rpc/__tests__/rpc-whitelist.unit.test.ts`
- `node scripts/npm-run.js npx react-doctor@latest . --verbose -y --project nextn --offline --diff main`

## Review Notes
- parallel subagent review found one actionable gap: missing route-level allowlist coverage for `ai_category_suggestion`; fixed in follow-up commit `67efe16`
- another subagent concern about lost `departmentList.equipment_count` was reviewed and rejected because `modelSummary.importantFields.departments` already preserves the counts in the landed code

---

<a href="https://app.devin.ai/review/thienchi2109/qltbyt-nam-phong/pull/155" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

---
## Summary by cubic
Completes Batch 2 payload compaction by moving `categorySuggestion` to a top‑k RPC and `departmentList` to the pass‑1 envelope, and by updating the system prompt to v2.5.0 with envelope-reading guidance and stricter tool-calling rules.

- **New Features**
  - Category suggestion: new `ai_category_suggestion` (top‑k) requiring `device_name`; registry maps to `p_device_name`; envelope exposes `candidates` in `modelSummary.importantFields`; SQL adds tenant scoping and FTS+trigram ranking (bounded).
  - Department list: pass‑1 envelope without `uiArtifact`; names and counts kept in `modelSummary.importantFields.departments`.
  - Prompt v2.5.0: use `modelSummary.itemCount` for counts and `uiArtifact.rawPayload.data` for current‑turn lists; ask for `device_name` before `categorySuggestion`; don’t reference `uiArtifact` in answers.

- **Refactors**
  - RPC proxy: whitelist includes `ai_category_suggestion` and `ai_department_list`; stricter session/claim typing, safer `rawBody` handling, and clearer error messages.
  - Tool metadata/UI: added `modelBudget` limits for `categorySuggestion` (10) and `departmentList` (50); added Vietnamese display names for both tools.

<sup>Written for commit 6352fe32e17ddac7263f97e3406cecae64372eee. Summary will update on new commits.</sup>

<h3>Greptile Summary</h3>

This PR completes Batch 2 of the `refactor-assistant-tool-payload-compaction` refactor by migrating `categorySuggestion` to a new bounded RPC (`ai_category_suggestion`) and moving `departmentList` to the pass-1 envelope contract (no `uiArtifact`). The system prompt is bumped to v2.5.0 with explicit envelope-reading guidance so the model no longer relies on the legacy `{ data, total }` shape.

Key changes:
- `categorySuggestion` now requires `device_name` (Zod `.strict()` schema) and maps it to `p_device_name` in the RPC call; the old open `{}` schema is replaced.
- `wrapRpcResultAsEnvelope` conditionally omits `uiArtifact` for `departmentList` and populates `modelSummary.importantFields` for both migrated tools via the new `buildImportantFields` helper.
- The SQL migration adds `ai_category_suggestion` with SECURITY DEFINER + pinned `search_path`, JWT guards, tenant isolation, FTS + trigram ranking, and bounded top-k candidate output. The previously flagged deduplication gap is correctly resolved by introducing `all_fts_matches` (unlimited) as the exclusion source for `trigram_matches`.
- Route-level cleanup in `rpc/[fn]/route.ts` removes all remaining `any` casts via the new `RpcProxySessionUser` type and helper functions, and tightens body parsing to guard against JSON arrays being spread.
- Test coverage is thorough: envelope shape, schema enforcement, RPC arg translation, and whitelist presence are all verified.

<h3>Confidence Score: 5/5</h3>

Safe to merge — no critical logic errors or security issues found; all prior review concerns are resolved.

The implementation is internally consistent: the Zod schema, RPC arg mapping, SQL function signature, whitelist, system prompt, and tests all agree on `device_name` → `p_device_name`. The FTS deduplication gap raised in the previous round is fixed. The route refactor removes `any` casts cleanly and the body-parsing change (guarding against arrays) is a strict improvement. Test coverage is comprehensive for the two migrated tools and the updated prompt section. No regressions were identified.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/app/api/rpc/[fn]/route.ts | Adds `ai_category_suggestion` to the whitelist (replacing `ai_category_list`), introduces `RpcProxySessionUser` type and helper functions to eliminate all remaining `any` casts, tightens body/rawBody typing to guard against arrays, and narrows the JWT claims type. |
| src/lib/ai/tools/registry.ts | Migrates `categorySuggestion` from `ai_category_list` (no input) to `ai_category_suggestion` (requires `device_name` via Zod `.strict()` schema). Adds `modelBudget` metadata to both tools and a parameter-mapping branch translating `device_name` → `p_device_name`. |
| src/lib/ai/tools/rpc-tool-executor.ts | Adds `buildImportantFields` for tool-specific important-field extraction and updates `wrapRpcResultAsEnvelope` to conditionally omit `uiArtifact` for `departmentList`. |
| supabase/migrations/20260328113000_add_ai_category_suggestion_rpc.sql | New migration adds `ai_category_suggestion` with SECURITY DEFINER, pinned search_path, JWT guards, tenant isolation, FTS+trigram ranking, and bounded top-k output. Deduplication uses `all_fts_matches` (no limit) — the prior review concern is resolved. |
| src/lib/ai/prompts/system.ts | Bumped to v2.5.0. Adds envelope-reading guidance block and updates `categorySuggestion` workflow to require `device_name` before invocation; removes references to the deprecated `{ data, total }` contract. |
| src/lib/ai/tools/__tests__/department-list-envelope.test.ts | New test file covering the `departmentList` pass-1 envelope: validates `uiArtifact` absence, `importantFields.departments` content, and edge cases. |
| src/lib/ai/tools/__tests__/category-suggestion-envelope.test.ts | New test file covering the `categorySuggestion` envelope: RPC mapping, schema enforcement, arg translation, envelope shape, and compaction behavior. |
| src/components/assistant/AssistantToolExecutionCard.tsx | Adds Vietnamese display names for `categorySuggestion` and `departmentList` to the `TOOL_DISPLAY_NAMES` map. |

</details>

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/lib/ai/tools/__tests__/department-list-envelope.test.ts`, line 803-822 ([link](https://github.com/thienchi2109/qltbyt-nam-phong/blob/67efe1664cb43e419ed1d60df53927f48510a76d/src/lib/ai/tools/__tests__/department-list-envelope.test.ts#L803-L822)) 

   <a href="#">P2</a> **Misleading test title — empty strings are kept, not filtered**

   The test is titled `'filters out non-string department names'` and the comment says *"Only rows with a string name should be included"*, but the expected output retains `{ name: '', equipment_count: 1 }`. An empty string is a valid `string`, so `typeof row.name === 'string'` is `true` for it. The implementation is consistent with the expectation, but the title and comment imply that empty names are rejected, which could confuse future maintainers. If empty strings should be excluded in practice, add `.min(1)` to the filter; otherwise rename the test and update the comment to reflect that the filter is `typeof name === 'string'` (inclusive of `''`).

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/ai/tools/__tests__/department-list-envelope.test.ts
   Line: 803-822

   Comment:
   **Misleading test title — empty strings are kept, not filtered**

   The test is titled `'filters out non-string department names'` and the comment says *"Only rows with a string name should be included"*, but the expected output retains `{ name: '', equipment_count: 1 }`. An empty string is a valid `string`, so `typeof row.name === 'string'` is `true` for it. The implementation is consistent with the expectation, but the title and comment imply that empty names are rejected, which could confuse future maintainers. If empty strings should be excluded in practice, add `.min(1)` to the filter; otherwise rename the test and update the comment to reflect that the filter is `typeof name === 'string'` (inclusive of `''`).

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<sub>Reviews (3): Last reviewed commit: ["test: clarify department list empty-stri..."](https://github.com/thienchi2109/qltbyt-nam-phong/commit/6352fe32e17ddac7263f97e3406cecae64372eee) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26640585)</sub>

### CODE REVIEW

File: src/app/api/rpc/[fn]/route.ts [hunk 1]
```diff
@@ -207,6 +214,24 @@ function sanitizeForLog(obj: unknown, depth = 0): unknown {
...
+  const user = session.user
+  if (!user || typeof user !== 'object') {
+    return null
+  }
+  return user as RpcProxySessionUser
+}
+
+function getErrorMessage(error: unknown): string {
+  return error instanceof Error ? error.message : 'Unexpected error'
+}
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply src/app/api/rpc/[fn]/route.ts]: Checked against the pre-refactor file at 672ab16. The 500 response body already returned `Unexpected error` before this helper existed; only the log fallback used `Unknown error`. I am leaving this unchanged because there is no API-response regression to fix here.
